server/src/main/java/io/unitycatalog/server/UnityCatalogServer.java

```diff
@@ -43,68 +46,101 @@
     Configurator.initialize(null, "etc/conf/server.log4j2.properties");
   }
 
-  Server server;
-  private static final String basePath = "/api/2.1/unity-catalog/";
-  private static final String controlPath = "/api/1.0/unity-control/";
-
   public UnityCatalogServer() {
-    this(8080);
+    this(new Builder());
   }
 
-  public UnityCatalogServer(int port) {
-
+  private UnityCatalogServer(Builder builder) {
+    setDefaults(builder);
     Path configurationFolder = Path.of("etc", "conf");
+    SecurityConfiguration securityConfiguration = new SecurityConfiguration(configurationFolder);
 
-    securityConfiguration = new SecurityConfiguration(configurationFolder);
-    securityContext =
+    this.securityContext =
         new SecurityContext(configurationFolder, securityConfiguration, "server", INTERNAL);
+    this.serverProperties = builder.serverProperties;
+    this.server = initializeServer(builder);
+  }
+
+  private void setDefaults(Builder builder) {
+    if (builder.port == 0) {
+      builder.port(8080);
+    }
+    if (builder.serverProperties == null) {
+      builder.serverProperties(ServerProperties.getInstance());
+    }
+    if (builder.credentialOperations == null) {
+      AwsCredentialVendor awsCredentialVendor = new AwsCredentialVendor(builder.serverProperties);
+      AzureCredentialVendor azureCredentialVendor =
+          new AzureCredentialVendor(builder.serverProperties);
+      GcpCredentialVendor gcpCredentialVendor = new GcpCredentialVendor(builder.serverProperties);
+      CredentialOperations credentialOperations =
+          new CredentialOperations(awsCredentialVendor, azureCredentialVendor, gcpCredentialVendor);
+      builder.credentialOperations(credentialOperations);
+    }
+  }
 
-    ServerBuilder sb = Server.builder().serviceUnder("/docs", new DocService()).http(port);
-    addServices(sb);
+  private Server initializeServer(Builder unityCatalogServerBuilder) {
+    ServerBuilder armeriaServerBuilder =
+        Server.builder()
+            .http(unityCatalogServerBuilder.port)
+            .serviceUnder("/docs", new DocService());
 
-    server = sb.build();
+    MetastoreRepository.getInstance().initMetastoreIfNeeded();
+    UnityCatalogAuthorizer authorizer =
+        initializeAuthorizer(unityCatalogServerBuilder.serverProperties);
+    addApiServices(armeriaServerBuilder, unityCatalogServerBuilder, authorizer);
+    addSecurityDecorators(
+        armeriaServerBuilder, unityCatalogServerBuilder.serverProperties, authorizer);
+
+    return armeriaServerBuilder.build();
   }
 
-  private void addServices(ServerBuilder sb) {
-    ObjectMapper unityMapper =
-        JsonMapper.builder().disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES).build();
-    JacksonRequestConverterFunction unityConverterFunction =
-        new JacksonRequestConverterFunction(unityMapper);
+  private JacksonRequestConverterFunction createRequestConverterFunction() {
```

Comment:
do you need these two functions in multiple places? If not, why create a separate function and not define it inline in the addApiServices